### PR TITLE
[SE-0155] Default Arguments In Enum Cases

### PR DESCRIPTION
**Pull request by CodaFi**
This patch is built on #20950 and has the same goals as #20496 but is quite a bit cleaner.

Posting this so I can get feedback on it and hopefully land this and the declcontext changes together.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by slavapestov** [comment]
@CodaFi now do subscript default arguments :)

File: lib/SILGen/RValue.cpp [hunk 1]
```diff
@@ -318,13 +318,8 @@ static void copyOrInitValuesInto(Initialization *init,
                 KIND == ImplodeKind::Copy, "Not handled by init");
   bool isInit = (KIND == ImplodeKind::Forward);
 
-  // First, unwrap one-element tuples, since we cannot lower them.
```
[the remark this answers is not in the document]
Comment (CodaFi) [reply lib/SILGen/RValue.cpp]: At one point.  I'll dump it.

File: lib/SILGen/SILGenApply.cpp [hunk 2]
```diff
@@ -5228,11 +5237,56 @@ void SILGenFunction::emitRawYield(SILLocation loc,
...
+  auto substLoweredType = destTL.getLoweredType().getASTType();
+  RValue(SGF, element, substLoweredType, outputValue).forwardInto(SGF, element, dest);
+}
+
 /// Emits SIL instructions to create an enum value. Attempts to avoid
 /// unnecessary copies by emitting the payload directly into the enum
 /// payload, or into the box in the case of an indirect payload.
 ManagedValue SILGenFunction::emitInjectEnum(SILLocation loc,
-                                            ArgumentSource payload,
+                                            RValue &&payload,
```
[the remark this answers is not in the document]
Comment (CodaFi) [reply lib/SILGen/SILGenApply.cpp]: Mostly nativity.  My impression is I can evaluate an ArgumentSource to an RValue, which I can parlay into another ArgumentSource and forward or to PreparedArguments which I can't (AFAICT).  FWIW, a lot of the supporting code is doing ArgumentSource-y things anyways so I'm not sure what optimizations I'm defeating here.

File: lib/SILGen/SILGenApply.cpp [hunk 3]
```diff
@@ -4498,6 +4500,9 @@ CallEmission::applyEnumElementConstructor(SGFContext C) {
   // construction.
   EnumElementDecl *element = callee.getEnumElementDecl();
 
+  // Emit default arguments now.
+  SGF.SGM.emitEnumDefaultArgGenerators(element, element->getParameterList());
```
[the remark this answers is not in the document]
Comment (CodaFi) [reply lib/SILGen/SILGenApply.cpp]: Yikes.  This belongs as part of the SGM visitor.

File: lib/SILGen/RValue.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (CodaFi) [reply lib/SILGen/RValue.cpp]: Now I remember: This is needlessly restrictive for single-element tuples which can and do show up all the time for enum associated values.  The test case I added exercises a crash that is caused by keeping this code in.

[the remark this answers is not in the document]
Comment (slavapestov) [reply lib/SILGen/RValue.cpp]: I found a bug that was fixed by removing this check just now, and I don't see any other tests failing without it so I'm going to remove it. Please wait until I land my change first, it adds the test case and I'm picking it over to 5.0 too.

[the remark this answers is not in the document]
Comment (slavapestov) [reply lib/SILGen/RValue.cpp]: https://github.com/apple/swift/pull/21394

[the remark this answers is not in the document]
Comment (CodaFi) [reply lib/SILGen/RValue.cpp]: Thanks Slava.  I’ll rebase in the morning

File: lib/SILGen/SILGenApply.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (CodaFi) [reply lib/SILGen/SILGenApply.cpp]: Actually, aren't the emit-into opts being defeated by the fake parameter emission ginning up parameters at +1?

**Comment by CodaFi** [comment]
@rjmccall Confirmed, those extra copies are the parameters values being emitted at +1.  Naively flipping the pseudo calling convention to +0 isn't necessarily the right answer here either.  Got any ideas?

**Comment by rjmccall** [comment]
Well, when I was looking at this briefly a few days ago (when deciding whether we could reasonably support varargs emission in 5.0, to which the answer was definitely "no"), I had two ideas:

- First, we already have an `ArgSpecialDest` concept which is used to thread argument values out to somewhere that isn't just the argument list.  (It's currently used for varargs, so that the arguments get forwarded into an array.)  It seems reasonable to take advantage of this for enum payloads.

- Second, I keep hoping that we can massively simplify this code by rethinking the representation of arguments in the AST.  If we can change `ApplyExpr` to take a list of arguments instead of an arbitrary tuple, and then `CSApply` can rewrite that list of arguments to be a *semantic* list of arguments where individual components can either be (1) an `Expr*` or (2) a reference to a default-argument emitter, and variadics are handled by introducing a `VarargExpr` that just collects all the component expressions/expansions as its children, then a ton of complexity disappears from SILGen.  Now, in the medium term, we do have to support constructing enums with tuple values, but (1) it's fine if that path doesn't get all the optimizations of the standard path, and (2) it's fine if we don't support default arguments and varargs on that path.

**Comment by slavapestov** [comment]
> Second, I keep hoping that we can massively simplify this code by rethinking the representation of arguments in the AST.

You and me both!

**Comment by CodaFi** [comment]
@rjmccall @slavapestov What remains before this can get merged?

**Comment by CodaFi** [comment]
@rjmccall @slavapestov What remains before this can get merged?

File: lib/SILGen/SILGenApply.cpp [hunk 4]
```diff
@@ -5243,11 +5249,56 @@ void SILGenFunction::emitRawYield(SILLocation loc,
...
+                                                     element, origFormalType,
+                                                     std::move(payload));
+
+  if (outputValue.isInContext()) return;
+
+  // Use RValue's forward-into-initialization code.  We have to lie to
+  // RValue about the formal type (by using the lowered type) because
+  // we're emitting into an abstracted value, which RValue doesn't
+  // really handle.
+  auto substLoweredType = destTL.getLoweredType().getASTType();
```
[the remark this answers is not in the document]
Comment (CodaFi) [reply lib/SILGen/SILGenApply.cpp]: I'll fix that part in a follow-up.  For now, I've updated the patch to just forward an `ArgumentSource` using the old machinery - which can still benefit from this kind of refactoring later.

**Comment by slavapestov** [comment]
Does this fix variadic parameters in enum cases too?

**Comment by CodaFi** [comment]
@rjmccall That should do it.

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by slavapestov** [comment]
@CodaFi I see. It would suffice to strip off the variadic bit and turn it into an array at that point, but it might require adjustment in a few other places as well. Alternatively we could allow lowering tuple types containing varargs again (they're represented exactly like arrays), but I would prefer if we didn't so that one day we can remove the flags field from TupleTypeElt altogether.

**Comment by CodaFi** [comment]
@swift-ci please test macOS platform

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
Alright, that should do it.

⛵️

**Comment by DevAndArtist** [comment]
@CodaFi shouldn't this be noted in the changelog, since it's kind of a new feature?!

**Comment by CodaFi** [comment]
It should, and it will be, but I’m not quite finished with what all I have planned in this area.

**Comment by DevAndArtist** [comment]
@CodaFi okay then, I really appreciate your hard work. 🎉 I hope I can catch up one day and wrap my head around the compiler as well and contribute as well (starting learning c++11 atm.)

**Pull request merged**